zipline/utils/calendars/trading_schedule.py

```diff
@@ -0,0 +1,379 @@
+            previous_open_and_close,
+            open_and_close_hook=self.start_and_end,
+            previous_scheduled_day_hook=self.previous_execution_day,
+        )
+        self.execution_day_distance = partial(
+            scheduled_day_distance,
+            all_days=self.all_execution_days,
+        )
+        self.execution_minutes_for_day = partial(
+            minutes_for_day,
+            open_and_close_hook=self.start_and_end,
+        )
+        self.execution_days_in_range = partial(
+            days_in_range,
+            all_days=self.all_execution_days,
+        )
+        self.execution_minutes_for_days_in_range = partial(
+            minutes_for_days_in_range,
+            days_in_range_hook=self.execution_days_in_range,
+            minutes_for_day_hook=self.execution_minutes_for_day,
+        )
+        self.add_execution_days = partial(
+            add_scheduled_days,
+            next_scheduled_day_hook=self.next_execution_day,
+            previous_scheduled_day_hook=self.previous_execution_day,
+            all_trading_days=self.all_execution_days,
+        )
+        self.next_execution_minute = partial(
+            next_scheduled_minute,
+            is_scheduled_day_hook=self.is_executing_on_day,
+            open_and_close_hook=self.start_and_end,
+            next_open_and_close_hook=self.next_start_and_end,
+        )
+        self.previous_execution_minute = partial(
+            previous_scheduled_minute,
+            is_scheduled_day_hook=self.is_executing_on_day,
+            open_and_close_hook=self.start_and_end,
+            previous_open_and_close_hook=self.previous_start_and_end,
+        )
+
+    @abstractproperty
+    def day(self):
+        """
+        A CustomBusinessDay defining those days on which the algorithm is
+        usually trading.
+        """
+        raise NotImplementedError()
+
+    @abstractproperty
+    def tz(self):
+        """
+        The native timezone for this TradingSchedule.
+        """
+        raise NotImplementedError()
+
+    @abstractproperty
+    def first_execution_day(self):
+        """
+        The first possible day of trading in this TradingSchedule.
+        """
+        raise NotImplementedError()
+
+    @abstractproperty
+    def last_execution_day(self):
+        """
+        The last possible day of trading in this TradingSchedule.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def trading_sessions(self, start, end):
+        """
+        Calculates all of the trading sessions between the given
+        start and end.
+
+        Parameters
+        ----------
+        start : Timestamp
+        end : Timestamp
+
+        Returns
+        -------
+        DataFrame
+            A DataFrame, with a DatetimeIndex of trading dates, containing
+            columns of trading starts and ends in this TradingSchedule.
+        """
+        raise NotImplementedError()
+
+    @property
+    @remember_last
```

Comment:
the cache will be invalidated if a subclass accesses it, so it is slightly different, though I think that is not intended.